### PR TITLE
Testing chapter

### PR DESCRIPTION
**Pull request by gadjevb**
The first two tasks for the testing chapter.

### CODE REVIEW

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package com.clouway.testing.test.task2;
+
+import com.clouway.testing.task2.ArrayMethods;
+import static org.junit.Assert.*;
+import org.junit.*;
+
+public class ArrayMethodsTest {
+    int arr[];
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: This is a smell and not required references. I have to lookup into setup method to find out what is my test data.

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        assertEquals("5 1 9 2 6 3 4 8 7 0 ",testTool.printArray(arr));
+    }
+
+    @Test(timeout = 1000)
+    public void sortArrayTest(){
+        testTool.sortArray(arr, 0, arr.length - 1);
+        assertArrayEquals(sortedArr, arr);
+    }
+
+    @Test(timeout = 1000)
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: timeout for such tasks is not a good option

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    public void reverseArrayTest(){
+        testTool.reverseArray(arr);
+        assertArrayEquals(reversedArr, arr);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void arrayNullPointerExceptionTest(){
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: You are adding Test as part of the name which is not necessary. 

Also arrayNullPointerException doesn't give good information. 

```java
@Test(expected = IllegalArgumentException.class)
public void getMinElementOfNullArray() {
 assertThat(new ArrayMethods().getMinElement(null));
}
```

As you can see It's better IllegalArgumentException to be thrown instead of NPE cause it gives you information that developer made a mistake and passed null as argument.

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test(timeout = 1000)
+    public void getSumOfArrayElementsTest(){
+        assertEquals(45,testTool.getSumOfArrayElements(arr));
+    }
+
+    @Test(timeout = 1000)
+    public void printArrayTest(){
+        assertEquals("5 1 9 2 6 3 4 8 7 0 ",testTool.printArray(arr));
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: This is not printing as it's rendering, i.e elements are rendered as space separated numbers. 

```java
String renderArray(int[] array); 
// you are asking someone to make transformation of the array 
// after that the caller need to do something with these numbers 
// For example: to print them 
```
vs

```java
// delegates to someone else to print the numbers and you don't care how this is going to happen 
void printArray(arr);
```

You have to find out how to test System.out invocations no matter that it's static variable. This will help you differentiate between testing of global state and local state.

File: src/com/clouway/testing/test/task1/SumatorTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public void numberFormatExceptionTest(){
+        addition.sumStrings("a12","11");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void illegalArgumentExceptionTest(){
+        addition.sumStrings(null,"3");
+    }
+
+    @After
```
Comment (mgenov) [inline src/com/clouway/testing/test/task1/SumatorTest.java]: The teardown method is not required.

File: src/com/clouway/testing/task2/ArrayMethods.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    public int getSumOfArrayElements(int arr[]) throws NullPointerException { // returns the sum of the elements in the array
+        int sum = 0;
+        for(int i = 0; i < arr.length; i++){
+            sum = sum + arr[i];
+        }
+        return sum;
+    }
+
+    public String printArray(int arr[]) throws NullPointerException { // prints all elements of the array
```
Comment (mgenov) [inline src/com/clouway/testing/task2/ArrayMethods.java]: It just formats them as separated string value and is not printing them.

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    public void arrayNullPointerExceptionTest(){
+        testTool.getMinElement(emptyArr);
+    }
+
+    @Test(expected = IndexOutOfBoundsException.class)
+    public void arrayIndexOutOfBoundsExceptionTest(){
+        testTool.getElement(arr, 15);
+    }
+
+    @After
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: This step is not required actually.

File: src/com/clouway/testing/task2/ArrayMethods.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        int temp = partition(arr,left,right);
+        if(left < temp - 1){
+            sortArray(arr,left,temp - 1);
+        }
+        if(temp < right) {
+            sortArray(arr, temp, right);
+        }
+    }
+
+    public int partition(int mas[], int left, int right) throws NullPointerException { // partition and sorting in quickSort
```
Comment (mgenov) [inline src/com/clouway/testing/task2/ArrayMethods.java]: Should not be with public visibility I suppose.

File: src/com/clouway/testing/task2/ArrayMethods.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    public String printArray(int arr[]) throws NullPointerException { // prints all elements of the array
+        String temp = "";
+        for(int i = 0; i < arr.length; i++){
+            temp = temp + arr[i] + " ";
+        }
+        return temp;
+    }
+
+    public void sortArray(int arr[], int left, int right) throws NullPointerException {
```
Comment (mgenov) [inline src/com/clouway/testing/task2/ArrayMethods.java]: Should not be with public visibility I suppose.

File: src/com/clouway/testing/task2/ArrayMethods.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
+package com.clouway.testing.task2;
+
+public class ArrayMethods {
+
+    public int getElement(int arr[], int position) throws IndexOutOfBoundsException {
```
Comment (mgenov) [inline src/com/clouway/testing/task2/ArrayMethods.java]: I'm not sure that it's part of the task.

**Comment by gadjevb** [comment]
I've removed to methods that are not part of the task. I have a question, why using (timeout = 1000) was a bad idea? I'm testing loops, if there is a mistake in the code in them they may end up to be endless. Even if they are not endless it would be good to have a time limit, a user won't use my code if he has to wait for one minute to give him the result.

**Comment by mgenov** [comment]
Because it leads to flaky tests on slow machines, t.e test sometimes passes, sometimes not, depending on the load of the machine.

When developers run tests continuously they will be able to figure that out on there own.

**Comment by gadjevb** [comment]
I'm not sure if the way I check for space in the Warehouse is good, there's no check for space for each type of product. I couldn't find a way that I'm going to foresee what products may be added to the warehouse.

File: src/com/clouway/testing/task2/ArrayMethods.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        return sum;
+    }
+
+    public void printArray(int arr[]) throws NullPointerException { // prints all elements of the array
+        for(int i = 0; i < arr.length; i++){
+            System.out.print(arr[i] + " ");
+        }
+    }
+
+    public void sortArray(int arr[], int left, int right) throws NullPointerException {
```
Comment (mgenov) [inline src/com/clouway/testing/task2/ArrayMethods.java]: left and right should not be visible for the clients of this method. 

If they are visible, I'm able to call it like this: sortArray(new int[] {3,1,23,12,45,33,11,10,9,34}, -1, 20);

File: src/com/clouway/testing/task3/Warehouse.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    private int spaceForMerchandise = 2;
+    ArrayList<Merchandise> stock = new ArrayList<>();
+
+    /*
+    addMerchandise() methods adds the given merchandise to the stock list after a check for space in the Warehouse
+    If spaceForMerchandise is greater than 0, the merchandise will be saved and spaceForMerchandise will be decremented,
+    otherwise WarehouseIsFullException will be thrown.
+     */
+
+    public void addMerchandise(Merchandise object) throws WarehouseIsFullException {
```
Comment (mgenov) [inline src/com/clouway/testing/task3/Warehouse.java]: Method comments should start with ** which is already described at: https://github.com/clouway/teamspace/wiki/Introducing%20Java in section "Документиране на кода с JavaDoc"

```java
/**
 * Adds the given merchandise to the stock list  if Warehouse contains space for it. 
 */ 
```

File: src/com/clouway/testing/test/task2/ArrayMethodsTest.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        int arr[] = {5,1,9,2,6,3,4,8,7,0};
+        assertEquals(45,testTool.getSumOfArrayElements(arr));
+    }
+
+    @Test
+    public void printArray(){
+        int arr[] = {5,1,9,2,6,3,4,8,7,0};
+        ArrayMethods testTool = new ArrayMethods();
+        ByteArrayOutputStream output = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(output));
```
Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: You are not restoring the original state after test is executed and test will leak ByteArrayOutputStream to others.

Comment (mgenov) [inline src/com/clouway/testing/test/task2/ArrayMethodsTest.java]: I'm talking about something like: 

```java
OutputStream originalSystemOut = System.out; 
try {
 // do testing stuff 
 System.setOut(new PrintStream(output);
} finally {
  System.setOut(originalSystemOut);
}
```

File: src/com/clouway/testing/test/task3/WarehouseTDD.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+public class WarehouseTDD {
+
+    /**
+    The first two test check is the Warehouse and Merchandise objects
+     are created.
+     */
+
+    @Test
+    public void createWarehouseObject(){
```
Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: You don't need this test actually cause java may never return null after object is instantiated. 

new XObject is never returning null value.

File: src/com/clouway/testing/test/task3/WarehouseTDD.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        Merchandise sortedTemp = sortedWarehouse.getMerchandiseAtPosition(0);
+        assertEquals(temp.getType(), sortedTemp.getType());
+        assertEquals(temp.getModel(), sortedTemp.getModel());
+        assertEquals(temp.getPrice(), sortedTemp.getPrice());
+    }
+
+    @Test
+    public void sortEmptyWarehouseMerchandiseByPrice(){
+        Warehouse warehouse = new Warehouse();
+        warehouse.sortByPrice();
```
Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: I don't see how you are asserting this stuff. Maybe something like:


```java
// Given that warehouse contains 2 merchants 
warehouse.addMerchandise(new Merchandise("Fridge", "Bosch", 300));
warehouse.addMerchandise(new Merchandise("TV", "Neo", 100));

// When Goods/Products/Merchants list sorted by price is retrieved 
List<Merchandise> goods = warehouse.listMerchantsByPrice();

// Then the retried list should be sorted in ascending order 
assertThat(goods.size(), is(equalTo(2)); // verifies that all merchandises are returned 
assertThat(goods.get(0).getPrice(), is(equalTo(100)));
assertThat(goods.get(0).getName(), is(equalTo("TV")));
assertThat(goods.get(1).getPrice(), is(equalTo(300)));
assertThat(goods.get(1).getName(), is(equalTo("Fridge")));
```

Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: Keep in mind, that if you override equals/hashCode you will be able to simplify this test to: 
```java
assertThat(goods, is(equalTo(
     Lists.newArrayList(
           new Merchandise("TV", "Neo", 100)
           new Merchandise("Fridge", "Bosch", 300),
     )
);
```

i.e test looks much more readable.

Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: In my example I've used Lists which is part of Guava library. 

http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Lists.html

**Comment by gadjevb** [comment]
I will checkout and try the Guava library when I get home.

**Comment by gadjevb** [comment]
I've overrided the equals() method in the Merchandise class, also I've added the Guava library and put Lists to good use, just like it was shown to me.

File: src/com/clouway/testing/task3/Merchandise.java [hunk 16]
```diff
@@ -22,4 +22,39 @@ public String getModel(){
     public int getPrice(){
         return price;
     }
+
+    /**
+     * Overriding for the equals() method so that it can compare Merchandise objects.
+     */
+
+    @Override
+    public boolean equals(Object obj){
```
Comment (mgenov) [inline src/com/clouway/testing/task3/Merchandise.java]: IDEA has a future for generation of such code: https://www.jetbrains.com/help/idea/2016.1/generating-equals-and-hashcode.html

It's not always good idea code to be generated, but you can take a look it as example. equals/hashCode are looking slightly better when generated.

File: src/com/clouway/testing/task3/Warehouse.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
+package com.clouway.testing.task3;
+
+import java.util.*;
+
+public class Warehouse {
+    private int spaceForMerchandise = 2;
+    ArrayList<Merchandise> stock = new ArrayList<>();
```
Comment (mgenov) [inline src/com/clouway/testing/task3/Warehouse.java]: In most cases it's better to use upper level interface for type of the reference. 

```java
private final List<Merchandise> stock = new ArrayList<>();
```

By doing this definition, you are denying the Warehouse to access methods that could be specific to ArrayList implementation only.

File: src/com/clouway/testing/task3/Warehouse.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     *addMerchandise() methods adds the given merchandise to the stock list after a check for space in the Warehouse
```
Comment (mgenov) [inline src/com/clouway/testing/task3/Warehouse.java]: Comments should start with verb, not as a name of method. 
http://www.oracle.com/technetwork/articles/java/index-137868.html

```java
/**
 * Returns an Image object that can then be painted on the screen. 
 * The url argument must specify an absolute {@link URL}. The name
 * argument is a specifier that is relative to the url argument. 
 * <p>
 * This method always returns immediately, whether or not the 
 * image exists. When this applet attempts to draw the image on
 * the screen, the data will be loaded. The graphics primitives 
 * that draw the image will incrementally paint on the screen. 
 *
 * @param  url  an absolute URL giving the base location of the image
 * @param  name the location of the image, relative to the url argument
 * @return      the image at the specified URL
 * @see         Image
 */
 public Image getImage(URL url, String name) {
        try {
            return getImage(new URL(url, name));
        } catch (MalformedURLException e) {
            return null;
        }
 }
```

Keep in mind that I'm talking only for the structure not how long or short comment should be.

File: src/com/clouway/testing/test/task1/SumatorTest.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    @Test(expected = NumberFormatException.class)
+    public void numberFormatException(){
+        Sumator addition = new Sumator();
+        addition.sumStrings("a12","11");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void illegalArgumentException(){
```
Comment (mgenov) [inline src/com/clouway/testing/test/task1/SumatorTest.java]: You are messing result from the name of the case which you are testing. You are verifying the scenario when argument is null. 

```java
@Test
public void argumentsAreNull() {
}
```
or 

```java
@Test
public void firstArgumentIsNull() {
}
```
if you are verifying only the case when first argument is null.

File: src/com/clouway/testing/test/task3/WarehouseTDD.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+
+import com.google.common.collect.Lists;
+import org.junit.*;
+import java.util.*;
+
+public class WarehouseTDD {
```
Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: Test names should be like: 
The test of Warehouse is WarehouseTest 
or in some cases: 

SellProductsFromWarehouseTest if you are verifying only scenarios with selling of products from warehouse.

File: src/com/clouway/testing/test/task3/WarehouseTDD.java [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+                        new Merchandise("Televisor","Neo",430),
+                        new Merchandise("Fridge","Bosch",800)
+                )
+        )));
+    }
+
+    @Test
+    public void sortEmptyWarehouseMerchandiseByPrice(){
+        Warehouse warehouse = new Warehouse();
+        warehouse.listMerchandiseByPrice();
```
Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: Not testing anything

If you going to test sorting when warehouse is empty, then returned list should be empty. 

```java
assertThat(warehouse.listMerchandiseByPrice(), is(equalTo(Collections.emptyList()));
```

File: src/com/clouway/testing/test/task3/WarehouseTDD.java [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+
+public class WarehouseTDD {
+
+    /**
+     *The first two test check the addMerchandise() method for the Warehouse class
+     *and the WarehouseIsFullException
+     */
+
+    @Test
+    public void addMerchandiseToWarehouse() {
```
Comment (mgenov) [inline src/com/clouway/testing/test/task3/WarehouseTDD.java]: You don't need this test as selling is verifying that already. 

You have to write tests for behaviour and not for methods. 

```java
// Given that warehouse contains 20 Apples
warehouse.addProduct(new Product("Apples",20)); 
// When 10 apples are sold
warehouse.sellProduct("Apples", 10);
// Then inventory should have only 10 apples 
Inventory inventory = warehouse.inventory();
assertThat(inventory.get("Apples").getQuanity(), is(equalTo(10)));

```

As you can see my test uses 3 methods which are working in combination and are representing the behaviour of the warehouse. 
"When in Warehouse are registered 20 apples and 10 of them are sold, then Warehouse should contain 10 apples to sell."

**Comment by gadjevb** [comment]
addMerchandise() and sellMerchandise() now have the ability to work with the quantity of the products. Which provides for a better simulation of a real warehouse.

**Comment by mgenov** [comment]
Tomorrow I'll check how you TDD'ing this task.

**Comment by gadjevb** [comment]
It's still not ready for review!

**Comment by gadjevb** [comment]
I will begin with the second task.

**Comment by gadjevb** [comment]
All tasks from the Testing with mocks chapter are ready for review!

File: src/com/clouway/testing/test/task4/EndpointFilterTest.java [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        }});
+        assertFalse(endpointFilter.shouldFilter("user"));
+    }
+
+    /**
+     * Mocks several Endpoints, all of them Fail
+     */
+
+    @Test
+    public void filterSeveralEndpointsallOfThemReturnFalse(){
```
Comment (mgenov) [inline src/com/clouway/testing/test/task4/EndpointFilterTest.java]: allEndpointsAreNotMatching sounds better.

File: src/com/clouway/testing/test/task4/EndpointFilterTest.java [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+        }});
+        assertFalse(endpointFilter.shouldFilter("/user"));
+    }
+
+    /**
+     * Mocks several Endpoints, at least one of them must be True
+     */
+
+    @Test
+    public void filterSeveralEndpointsAtLeastOneWillReturnTrue(){
```
Comment (mgenov) [inline src/com/clouway/testing/test/task4/EndpointFilterTest.java]: If your test ends with "matching" then you stick to naming of the endpoint which is much more meaningful than "return true" cause many things may return true.

File: src/com/clouway/testing/test/task5/SMSValidatorTest.java [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    public void smsIsWithoutABody(){
+        assertFalse(validator.isSmsValid(new SMS("0878884433","Work","")));
+    }
+
+    @Test
+    public void textInSmsBodyIsWithIncorrectLength(){
+        String flag = veryLongString(101);
```
Comment (mgenov) [inline src/com/clouway/testing/test/task5/SMSValidatorTest.java]: I see that you are using this word flag on many places which doesn't sounds understandable. 

String longBodyMessage = veryLongString(101); 
vs 
String flag = veryLongString(101);

File: src/com/clouway/testing/test/task6/UserRepositoryTest.java [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+            oneOf(validator).isValid("21");
+            will(returnValue(true));
+
+            oneOf(userDB).Register(adultUser);
+            will(returnValue(true));
+
+            oneOf(userDB).getRegisteredUser("Georgi");
+            will(returnValue(false));
+        }});
+        userRepository.registerUser(adultUser);
```
Comment (mgenov) [inline src/com/clouway/testing/test/task6/UserRepositoryTest.java]: When you are testing in isolation it's not reasonable to call 2 methods which are using mocks, cause with mocks you are able to simulate any behaviour.

Comment (mgenov) [inline src/com/clouway/testing/test/task6/UserRepositoryTest.java]: If you want to test in this manner, then a fake object is more reasonable.

File: src/com/clouway/testing/test/task5/SMSValidatorTest.java [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Test
+    public void textInSmsBodyIsWithIncorrectLength(){
+        String longBodyMessage = veryLongString(101);
+        assertFalse(validator.isSmsValid(new SMS("0878884433","Work",longBodyMessage)));
+    }
+
+    @Test
+    public void textInSmsTitleIsWithIncorrectLength(){
+        String flag = veryLongString(31);
```
Comment (mgenov) [inline src/com/clouway/testing/test/task5/SMSValidatorTest.java]: .

**Comment by mgenov** [comment]
LGTM !
